### ISSUE TITLE
Error with recursive definition

### ISSUE THREAD

**Issue by coqbot** [coq/coq#3738]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#3738
From: @dasuxullebt
Reported version: 8.5
CC:   @dasuxullebt

**Comment by Alizter** [comment]
@maximedenes 

This is still an issue in 8.15+alpha.

I have resurrected and adapted the code and run the bug minimizer on it. Here is a minimal example:
<details>
<summary>Minimal example</summary>

```coq
(* -*- mode: coq; coq-prog-args: ("-emacs" "-w" "-deprecated-native-compiler-option" "-R" "." "Top" "-top" "InflateUnverified" "-native-compiler" "ondemand") -*- *)
(* File reduced by coq-bug-finder from original input, then from 258 lines to 146 lines, then from 160 lines to 194 lines, then from 198 lines to 153 lines, then from 157 lines to 153 lines *)
(* coqc version 8.15+alpha compiled with OCaml 4.12.0
   coqtop version 8.15+alpha *)
Require Coq.funind.Recdef.
Require Coq.Vectors.Vector.

Notation vec := Vector.t.
Notation LB := (list bool).

Notation Byte := (vec bool 8).

Notation LByte := (list Byte).

Import Coq.Lists.List.

Inductive BackBufferBase (A : Set) : Set :=
| bbnil : BackBufferBase A
| bbcons1 : A -> BackBufferBase (A * A)%type -> BackBufferBase A
| bbcons2 : A -> A -> BackBufferBase (A * A)%type -> BackBufferBase A.

Record BackBufferWithLength (A : Set) : Set :=
  mkBackBufferWithLength {
      l : nat ;
      buffer : BackBufferBase A
    }.

Fixpoint bbbcons {A : Set} (x : A) (bbb : BackBufferBase A) : BackBufferBase A :=
  match bbb with
    | bbnil _ => bbcons1 A x (bbnil (A * A)%type)
    | bbcons1 _ a bb => bbcons2 A x a bb
    | bbcons2 _ a b bb => bbcons1 A x (bbbcons (a, b) bb)
  end.

Function bbwlcons {A : Set} (x : A) (bbwl : BackBufferWithLength A) : BackBufferWithLength A :=
  mkBackBufferWithLength A (S (l A bbwl)) (bbbcons x (buffer A bbwl)).

Record BackBuffer (A : Set) : Set :=
  mkBackBuffer {
      length : nat ;
      topbuf : BackBufferWithLength A ;
      botbuf : BackBufferWithLength A
    }.

Function bbcons {A : Set} (x : A) (bb : BackBuffer A) : BackBuffer A :=
  match Compare_dec.le_dec (l A (topbuf A bb)) (length A bb) with
    | left _ => mkBackBuffer A (length A bb) (bbwlcons x (topbuf A bb)) (botbuf A bb)
    | right _ => mkBackBuffer
                   A
                   (length A bb)
                   (mkBackBufferWithLength A 1
                                           (bbcons1 A x (bbnil (A * A)%type)))
                   (topbuf A bb)
  end.

Inductive BitByteList : Set :=
  | BBLNil : BitByteList
  | BBLConsBit : bool -> BitByteList -> BitByteList
  | BBLConsByte : Byte -> BitByteList -> BitByteList -> BitByteList.

Fixpoint bbllength (bbl : BitByteList) : nat :=
  match bbl with
    | BBLNil => 0
    | BBLConsBit _ bb => S (bbllength bb)
    | BBLConsByte _ bb _ => bbllength bb
  end.

Fixpoint nextBit (BBL : BitByteList) : option (bool * BitByteList) :=
  match BBL with
    | BBLNil => None
    | BBLConsBit bit bb => Some (bit, bb)
    | BBLConsByte _ bb _ => nextBit bb
  end.

Fixpoint nBits (BBL : BitByteList) (n : nat) : option (LB * BitByteList) :=
  match n with
    | 0 => Some (nil, BBL)
    | (S n) => match nextBit BBL with
                 | None => None
                 | Some (bit, rest) =>
                   option_map (fun x => (bit :: fst x, snd x)) (nBits rest n)
               end
  end.

Fixpoint lsbToInt (l : LB) : nat :=
  match l with
    | nil => 0
    | false :: R => 2 * (lsbToInt R)
    | true :: R => S (2 * (lsbToInt R))
  end.

Definition StreamEndedEarly := 1.
Definition WrongBlockHeader := 2.

Inductive CoqCantIntoMutualRecursion : Set :=
| Start : CoqCantIntoMutualRecursion
| BlockParse : bool -> CoqCantIntoMutualRecursion
| UncompressedBlockParse : bool -> CoqCantIntoMutualRecursion
| UncompressedBlockPass : nat -> bool -> CoqCantIntoMutualRecursion
| DynamicallyCompressedBlockParse : bool -> CoqCantIntoMutualRecursion
| StaticallyCompressedBlockParse : bool -> CoqCantIntoMutualRecursion.

Function inflate (status : CoqCantIntoMutualRecursion) (input : BitByteList) (back : BackBuffer Byte)
         (onerror :  CoqCantIntoMutualRecursion -> BitByteList -> nat -> LByte) {measure bbllength input} : LByte :=
  match status with
    | Start => match (nextBit input) with
                 | None => onerror Start input StreamEndedEarly
                 | Some (bit, rest) => inflate (BlockParse bit) rest back onerror
               end
    | BlockParse x =>
      match (nBits input 2) with
        | None => onerror (BlockParse x) input StreamEndedEarly
        | Some (lb, rest) =>
          match lb with
            | false :: false :: nil => inflate (UncompressedBlockParse x) rest back onerror
            | false :: true :: nil => inflate (DynamicallyCompressedBlockParse x) rest back onerror
            | true :: false :: nil => inflate (StaticallyCompressedBlockParse x) rest back onerror
            | _ => onerror (BlockParse x) rest WrongBlockHeader
          end
      end
    | UncompressedBlockParse x =>
      match input with
        | BBLConsBit bit rest => inflate (UncompressedBlockParse x) rest back onerror
        | BBLConsByte _ bb _ =>
          match nBits bb 16 with
            | None => onerror (UncompressedBlockParse x) bb 0
            | Some (bits1, bbb) =>
              match nBits bbb 16 with
                | None => onerror (UncompressedBlockParse x) bb 1
                | Some (bits2, rest) =>
                  (fix f a b :=
                           match (a, b) with
                             | (true :: L1, true :: L2) => f L1 L2
                             | (false :: L1, false :: L2) => f L1 L2
                             | (nil, nil) => inflate (UncompressedBlockPass (lsbToInt bits1) x) rest back onerror
                             | _ => onerror (UncompressedBlockParse x) rest 2
                           end) bits1 bits2
              end
          end
        | _ => onerror (UncompressedBlockParse x) input 3
      end
    | UncompressedBlockPass n x =>
      match n with
        | 0 => match x with
                 | false => inflate Start input back onerror
                 | true => nil
               end
        | (S n) =>
          match input with
            | BBLConsByte byte _ bb => byte :: inflate (UncompressedBlockPass n x) bb (bbcons byte back) onerror
            | _ => onerror (UncompressedBlockPass n x) input 0
          end
      end
    | _ => onerror Start input 9
  end.

```
</details>

**Comment by mrhaandi** [comment]
A further minimized example:
```coq
Require Coq.funind.Recdef.

Function inflate (n : nat) {measure (fun x : nat => x) n} : unit :=
  (fix f a := match a with
              | S m => f m
              | _ => tt
              end) n.
(*
Anomaly
"travel_aux : unexpected (fix f (a : nat) : unit :=
                            match a with
                            | 0 => tt
                            | S m => f m
                            end) n."
*)
```

**Comment by herbelin** [comment]
Duplicate of #17110, fixed in #17113.
